### PR TITLE
Add CFG Rescale option for supporting zero-terminal SNR models

### PR DESCRIPTION
**Pull request by damian0815**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [ ] Yes
- [X] No, because:

      
## Have you updated all relevant documentation?
- [ ] Yes
- [X] No

## Description

This PR implements CFG rescaling as required when using models trained using zero-terminal SNR. It exposes a `CFG Rescale Multiplier` input on the Denoise Latents node:
Screenshot 2023-08-20 at 12 52 38

What's this for? It is required for models trained using zero-terminal SNR (ref https://arxiv.org/abs/2305.08891 -- the "proper" implementation of what is known elsewhere as "offset noise"), mainly to avoid issues with oversaturation. 
Here's an example from a model I trained using zero-terminal SNR. This is the output without CFG rescaling:
a4f604d6-2846-41d3-b416-0ab811520f43

and here's the same prompt/seed with CFG rescaling set to 0.7 (which is the recommended value for zero-terminal SNR trained models):
364cde7b-3905-4546-975f-650ffc17ff0d

Note the improved contrast and details.

### CODE REVIEW

**Comment by damian0815** [comment]
Note that there is an implementation of rescale CFG in upstream diffusers but a) according to @SargeZT it's not implemented correctly and b) Invoke's custom diffusers pipeline does not have ready access to it anyway

**Comment by psychedelicious** [comment]
Would users of these models know that this setting needs to be used? 

Is there a programmatic way to determine if a model needs the setting to be used?

File: invokeai/app/invocations/latent.py [hunk 1]
```diff
@@ -118,6 +118,9 @@ class DenoiseLatentsInvocation(BaseInvocation):
     )
     denoising_start: float = InputField(default=0.0, ge=0, le=1, description=FieldDescriptions.denoising_start)
     denoising_end: float = InputField(default=1.0, ge=0, le=1, description=FieldDescriptions.denoising_end)
+    cfg_rescale_multiplier: float = InputField(
+        default=0, ge=0, lt=1, description=FieldDescriptions.cfg_rescale_multiplier, ui_type=UIType.Float
+    )
```
Comment (psychedelicious) [inline invokeai/app/invocations/latent.py:123]: You can safely leave off the `ui_type=UIType.Float` on this field - the client will infer it as a float correctly

**Comment by keturn** [comment]
> Would users of these models know that this setting needs to be used?
> 
> Is there a programmatic way to determine if a model needs the setting to be used?

I was thinking along these lines too. If it's not something we expect to toggle on and off for each denoising operation, and it's instead something we always expect to apply to some models and never to others, then it sounds like we could model it as a property of the model.

It could go in its `unet/config.json` and we wouldn't need to clutter the interface with it.

I don't know, perhaps there's a case from altering it from that "recommended" value? In which case I suppose we could expose it on the Denoise Latents interface, but we'd want it to use its recommended value by default unless you specifically override it.

**Comment by damian0815** [comment]
it could probably be set on a per-model basis. what i'm unsure of is whether there's any point trying other values than `0.7` (ZTSNR models) or `0` (regular models)

File: invokeai/backend/stable_diffusion/diffusion/shared_invokeai_diffusion.py [hunk 2]
```diff
@@ -30,6 +30,18 @@
     Callable[[torch.Tensor, torch.Tensor, torch.Tensor], torch.Tensor],
 ]
 
+def rescale_noise_cfg(noise_cfg, noise_pred_text, guidance_rescale=0.0):
```
Comment (StAlKeR7779) [inline invokeai/backend/stable_diffusion/diffusion/shared_invokeai_diffusion.py:33]: Unused?

File: invokeai/backend/stable_diffusion/diffusers_pipeline.py [hunk 3]
```diff
@@ -533,8 +534,15 @@ def step(
         noise_pred = self.invokeai_diffuser._combine(
             uc_noise_pred,
             c_noise_pred,
-            guidance_scale,
+            guidance_scale
         )
+        guidance_rescale_multiplier = conditioning_data.guidance_rescale_multiplier
+        if guidance_rescale_multiplier > 0:
+            noise_pred = type(self).rescale_cfg(
```
Comment (StAlKeR7779) [inline invokeai/backend/stable_diffusion/diffusers_pipeline.py:541]: There no need to extract type, classmethods will work correctly being called on self too.
Also you can make it here staticmethod.
// staticmethods too can be called normally on self.

File: invokeai/app/invocations/baseinvocation.py [hunk 4]
```diff
@@ -32,6 +32,7 @@ class FieldDescriptions:
     denoising_start = "When to start denoising, expressed a percentage of total steps"
     denoising_end = "When to stop denoising, expressed a percentage of total steps"
     cfg_scale = "Classifier-Free Guidance scale"
+    cfg_rescale_multiplier = ("Rescale multiplier for CFG guidance. Set to 0.7 for models trained with zero-terminal SNR")
```
Comment (StAlKeR7779) [inline invokeai/app/invocations/baseinvocation.py:35]: Could you remove brackets? For consistency with other lines.

**Comment by ljleb** [comment]
In the paper that introduces CFG rescale phi, they don't give a good justification as to why 0.7 is the best value. It's just what "looked right" to them in their environment. AFAICT this lacks sufficient rigour for 0.7 to be taken as an objective optimal value for models that were trained with offset noise.

**Comment by damian0815** [comment]
right, so that means it should probably be exposed as an argument?

the issue is that the latents node is already pretty over-burdened with options. i guess what would be ideal is having some way of grouping inputs to a node and having the groups expandable/collapsible.

**Comment by ljleb** [comment]
To my understanding, while storing a CFG rescale with each model could work, personally I like to tweak it for each generation I make.

I'm not aware of all the plans for invoke, but if there was a way in the graph to create loops and control flow conditions, it would be possible to create a custom denoising loop graph using even more basic nodes. all the little extensions that have been discovered this last year for latent diffusion models would become possible without having to fit them all into one node. In fact, this would also enable use cases like frame by frame video processing.

In my opinion, in the long run, instead of adding new fields to the denoising nodes, I think there should be some effort to break it into smaller pieces.

I presume something like graph loops could take a while if it is ever done, so maybe having the CFG rescale as a model property or grouped field set to either 0 or 0.7, or something like that, is good enough for now.

File: invokeai/backend/stable_diffusion/diffusers_pipeline.py (hunk 3 above)
Comment (psychedelicious) [reply invokeai/backend/stable_diffusion/diffusers_pipeline.py]: done

File: invokeai/backend/stable_diffusion/diffusion/shared_invokeai_diffusion.py (hunk 2 above)
Comment (psychedelicious) [reply invokeai/backend/stable_diffusion/diffusion/shared_invokeai_diffusion.py]: its removed

File: invokeai/backend/stable_diffusion/diffusers_pipeline.py [hunk 5]
```diff
@@ -633,6 +640,15 @@ def step(
 
         return step_output
 
+    @classmethod
+    def _rescale_cfg(cls, total_noise_pred, pos_noise_pred, multiplier=0.7):
```
[the remark this answers is not in the document]
Comment (psychedelicious) [reply invokeai/backend/stable_diffusion/diffusers_pipeline.py]: done

**Review by psychedelicious — APPROVED** [review]
I've added the setting to the linear UI.

Also took the liberty of tidying up a few things and addressing some low-hanging-fruit feedback.

**Comment by psychedelicious** [comment]
@RyanJDick @StAlKeR7779 please re-review

**Review by RyanJDick — COMMENTED** [review]
Took another look and tested the UI without issues. Would still like to get an answer to my unresolved comment, but looks good otherwise.

**Comment by keturn** [comment]
Here's another such model, based on SD 1.5, that might work with this: https://huggingface.co/drhead/ZeroDiffusion

**Review by hipsterusername — APPROVED** [review]

File: invokeai/backend/stable_diffusion/diffusers_pipeline.py [hunk 6]
```diff
@@ -633,6 +640,15 @@ def step(
...
+        ro_pos = torch.std(pos_noise_pred, dim=(1, 2, 3), keepdim=True)
+        ro_cfg = torch.std(total_noise_pred, dim=(1, 2, 3), keepdim=True)
+
+        x_rescaled = total_noise_pred * (ro_pos / ro_cfg)
+        x_final = multiplier * x_rescaled + (1.0 - multiplier) * total_noise_pred
+        return x_final
```
[the remark this answers is not in the document]
Comment (damian0815) [reply invokeai/backend/stable_diffusion/diffusers_pipeline.py:647]: my implementation is a copy/paste of SargeZT's, who is sadly no longer with us. i think you're right that they're equivalent, though.

[the remark this answers is not in the document]
Comment (psychedelicious) [reply invokeai/backend/stable_diffusion/diffusers_pipeline.py:648]: I've created #5199 as a reminder to revisit this implementation.

**Comment by psychedelicious** [comment]
I've resolved a number of conflicts on this branch and created #5199 to followup on the remaining discussion point.

**Pull request merged**